File: framework/languages/java/e2e/RegistrationCodec/Server/CodecRequester/src/main/java/systems/zlink/e2e/registrationcodec/codecrequester/Configuration/CodecRequesterOptions.java
package systems.zlink.e2e.registrationcodec.codecrequester.Configuration;
import org.springframework.boot.context.properties.ConfigurationProperties;

@ConfigurationProperties("e2e")
public record CodecRequesterOptions(
    String serverEndpoint,
    String httpEndpoint,
    String logDir) {
    public CodecRequesterOptions {
        required(serverEndpoint, "server-endpoint"); required(httpEndpoint, "http-endpoint"); required(logDir, "log-dir");
    }
    private static void required(String value, String name) {
        if (value == null || value.isBlank()) throw new IllegalArgumentException("e2e." + name + " is required");
    }
}

File: framework/languages/java/e2e-kotlin/RegistrationCodec/Server/CodecRequester/src/main/kotlin/systems/zlink/e2e/kotlin/registrationcodec/codecrequester/Configuration/CodecRequesterOptions.kt
package systems.zlink.e2e.kotlin.registrationcodec.codecrequester.configuration

data class CodecRequesterOptions(
    val httpEndpoint: String,
    val serverEndpoint: String,
    val logDir: String,
) {
    companion object {
        fun parse(args: Array<String>): CodecRequesterOptions {
            val values = mutableMapOf<String, MutableList<String>>()
            var index = 0
            while (index < args.size) {
                val key = args[index]
                require(key.startsWith("--")) { "Unexpected argument '$key'." }
                require(index + 1 < args.size) { "Missing value for '$key'." }
                values.getOrPut(key) { mutableListOf() }.add(args[index + 1])
                index += 2
            }

            fun required(name: String): String =
                values[name]?.lastOrNull()?.takeIf { it.isNotBlank() }
                    ?: throw IllegalArgumentException("$name is required.")
            fun get(name: String): String? = values[name]?.lastOrNull()
            return CodecRequesterOptions(
                httpEndpoint = required("--http-endpoint"),
                serverEndpoint = required("--server-endpoint"),
                logDir = get("--log-dir") ?: "logs",
            )
        }
    }
}
